src/blob/index/core.rs

```diff
@@ -0,0 +1,290 @@
+            self.inner = State::OnDisk(findex);
+            self.mem = None;
+            Ok(size)
+        } else {
+            Ok(0)
+        }
+    }
+
+    async fn load_in_memory(&mut self, findex: FileIndex) -> Result<()> {
+        let (record_headers, records_count) = findex.get_records_headers().await?;
+        self.mem = Some(compute_mem_attrs(&record_headers, records_count));
+        self.inner = State::InMemory(record_headers);
+        self.filter = findex.read_filter().await?;
+        Ok(())
+    }
+
+    pub(crate) fn memory_used(&self) -> usize {
+        if let State::InMemory(data) = &self.inner {
+            let mem = self
+                .mem
+                .as_ref()
+                .expect("No memory info in `InMemory` State");
+            trace!("record_header_size: {}, records_allocated: {}, data.len(): {}, entry_size (key + vec): {}",
+            mem.record_header_size, mem.records_allocated, data.len(), mem.btree_entry_size
+            );
+            // last minus is neccessary, because allocated but not initialized record headers don't
+            // have key allocated on heap
+            mem.record_header_size * mem.records_allocated + data.len() * mem.btree_entry_size
+                - (mem.records_allocated - mem.records_count) * mem.key_size
+        } else {
+            0
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<FileIndex: FileIndexTrait + Sync + Send + Clone> IndexTrait for IndexStruct<FileIndex> {
+    async fn contains_key(&self, key: &[u8]) -> Result<bool> {
+        self.get_any(key).await.map(|h| h.is_some())
+    }
+
+    fn push(&mut self, h: RecordHeader) -> Result<()> {
+        debug!("blob index simple push");
+        match &mut self.inner {
+            State::InMemory(headers) => {
+                debug!("blob index simple push bloom filter add");
+                self.filter.add(h.key());
+                debug!("blob index simple push key: {:?}", h.key());
+                let mem = self
+                    .mem
+                    .as_mut()
+                    .expect("No memory info in `InMemory` State");
+                // Same reason to use get_mut as in deserialize_record_headers.
+                if let Some(v) = headers.get_mut(h.key()) {
+                    let old_capacity = v.capacity();
+                    v.push(h);
+                    trace!("capacity growth: {}", v.capacity() - old_capacity);
+                    mem.records_allocated += v.capacity() - old_capacity;
+                } else {
+                    if mem.records_count == 0 {
+                        // record header contains key as Vec<u8>, h.key().len() - data on the heap
+                        mem.record_header_size = size_of::<RecordHeader>() + h.key().len();
+                        // every entry also includes data on heap: capacity * size_of::<RecordHeader>()
+                        mem.key_size = h.key().len();
+                        mem.btree_entry_size =
```

Comment:
this size is calculated more then once with same code. check `compute_mem_attrs`